builder/builder.go

```diff
@@ -0,0 +1,569 @@
+
+func (e *expression) Var(v string) *expression {
+	if len(v) == 0 {
+		panic("variable name cannot be empty")
+	}
+
+	if v[0] != '$' {
+		panic(fmt.Sprintf("variable must start with $: %s", v))
+	}
+
+	e.operand = variable(v)
+
+	return e
+}
+
+type option struct { //nolint:govet
+	key     string
+	operand any // literal or variable
+}
+
+func Option(key string, operand any) option { return option{key, operand} }
+
+func (e *expression) Func(name string, option ...option) *expression {
+	if len(name) == 0 {
+		panic("function name cannot be empty")
+	}
+
+	if e.function.name != "" {
+		panic("expression already has a function")
+	}
+
+	switch name[0] {
+	default:
+		panic(fmt.Sprintf("function MUST start with :, + or -: %s", name))
+	case ':', '+', '-':
+		e.function = function{name: name, options: option}
+		return e
+	}
+}
+
+func printLiteral(l any) string {
+	switch v := l.(type) {
+	case string:
+		return "|" + quotedEscape(v) + "|"
+	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
+		return fmt.Sprintf("%d", v)
+	case float32, float64:
+		return fmt.Sprintf("%f", v)
+	default:
+		panic(fmt.Sprintf("unsupported literal type: %T", v))
+	}
+}
+
+// helpers
+
+/*
+	quotedEscape escapes special characters in quoted name literal.
+
+ABNF:
+quoted-escape   = backslash ( backslash / "|" )
+.
+*/
+func quotedEscape(s string) string {
+	var sb strings.Builder
+
+	for _, c := range s {
+		switch c {
+		case '\\', '|':
+			sb.WriteRune('\\')
+		}
+
+		sb.WriteRune(c)
+	}
+
+	return sb.String()
+}
+
+/*
+	textEscape escapes special characters in text.
+
+ABNF:
+text-escape     = backslash ( backslash / "{" / "}" )
+.
+*/
+func textEscape(s string) string {
+	var sb strings.Builder
+
+	for _, c := range s {
+		switch c {
+		case '\\', '{', '}':
+			sb.WriteRune('\\')
+		}
+
+		sb.WriteRune(c)
+	}
+
+	return sb.String()
+}
+
+// hasSimpleStart returns true if the string has a simple start.
+// ABNF:
+// simple-start = simple-start-char / text-escape / placeholder
+// .
+func hasSimpleStart(s string) bool {
+	if len(s) > 0 {
+		c := []rune(s)[0]
+
+		if isSimpleStart(c) || c == '{' || c == '}' || c == '\\' {
+			return true
+		}
+	}
+
+	return false
+}
+
+// isSimpleStart returns true if r is simple start character.
+//
+// ABNF:
+//
+//	simple-start-char = %x0-2D         ; omit .
+//	                  / %x2F-5B        ; omit \
+//	                  / %x5D-7A        ; omit {
+//	                  / %x7C           ; omit }
+//	                  / %x7E-D7FF      ; omit surrogates
+//	                  / %xE000-10FFFF
+func isSimpleStart(r rune) bool {
+	return 0x0 <= r && r <= 0x2D || // omit .
+		0x2F <= r && r <= 0x5B || // omit \
+		0x5D <= r && r <= 0x7A || // omit {}
+		r == 0x7C || // omit }
+		0x7E <= r && r <= 0xD7FF || // omit surrogates
+		0xE000 <= r && r <= 0x10FFFF
+}
+
+// IsEmpty returns true if the builder is empty.
+func (b Builder) IsEmpty() bool {
+	if b.locals == nil && b.inputs == nil && b.selectors == nil &&
```

Comment:
**Never** check if the slice is empty by comparing it to nil.

IT IS BAD, and causes issues.

Nil vs empty slices.